merkle/tree_hasher.go

```diff
@@ -16,60 +16,33 @@ package merkle
 
 import (
 	"crypto"
-)
+	"fmt"
 
-// TODO(al): investigate whether we need configurable TreeHashers for
-// different users. Apparently E2E hashes in tree-level to the internal nodes
-// for example, and some users may want different domain separation prefixes
-// etc.
-//
-// BIG SCARY COMMENT:
-//
-// We don't want this code to have to depend on or constrain implementations for
-// specific applications but we haven't decided how we're going to split domain
-// specific stuff from the generic yet and we don't want to lose track of the fact
-// that this hashing needs to be domain aware to some extent.
-//
-// END OF BIG SCARY COMMENT
+	"github.com/google/trillian/merkle/rfc6962"
+)
 
-// Domain separation prefixes
-// TODO(Martin2112): Move anything CT specific out of here to <handwave> look over there
 const (
-	RFC6962LeafHashPrefix = 0
-	RFC6962NodeHashPrefix = 1
+	// RFC6962SHA256Type is the string used to retrieve the RFC6962 hasher.
+	RFC6962SHA256Type = "RFC6962-SHA256"
 )
 
-// TreeHasher implements the RFC6962 tree hashing algorithm.
-type TreeHasher struct {
-	crypto.Hash
-}
-
-// NewRFC6962TreeHasher creates a new TreeHasher based on the passed in hash function.
-// TODO(Martin2112): Move anything CT specific out of here to <handwave> look over there
-func NewRFC6962TreeHasher() TreeHasher {
-	return TreeHasher{Hash: crypto.SHA256}
-}
-
-// HashEmpty returns the hash of an empty element for the tree
-func (t TreeHasher) HashEmpty() []byte {
-	return t.New().Sum(nil)
+// TreeHasher is the interface that the previous tree hasher struct implemented.
+type TreeHasher interface {
+	HashEmpty() []byte
+	HashLeaf(leaf []byte) []byte
+	HashChildren(l, r []byte) []byte
+	Size() int
```

Comment:
In a future PR we can rename this BitLength() and return the number of bits in the hash to as to remove many of the `Size()*8` bits I see in the code.